### PR TITLE
Baremetal API V1: Allocations

### PR DESCRIPTION
**Pull request by stbenjam**
This PR adds support for Ironic Allocations.

For #1429

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

- API Ref: https://developer.openstack.org/api-ref/baremetal/?expanded=create-allocation-detail,list-allocations-detail
- Base class: https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/allocation.py#L39-L70
- List https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/allocation.py#L274-L298
-  Get https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/allocation.py#L306-L318
- Create https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/allocation.py#L320-L378

### CODE REVIEW

**Comment by stbenjam** [comment]
This should be ready for a review, thanks.

**Comment by stbenjam** [comment]
@dtantsur FYI

File: acceptance/openstack/baremetal/v1/allocations_test.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// +build acceptance baremetal allocations
+
+package v1
+
+import (
+	"testing"
+
+	"github.com/gophercloud/gophercloud/acceptance/clients"
+	v1 "github.com/gophercloud/gophercloud/acceptance/openstack/baremetal/v1"
```
Comment (jtopjian) [inline acceptance/openstack/baremetal/v1/allocations_test.go]: This line can be removed - it's importing itself. I think this is what's causing the OpenLab failure.

File: openstack/baremetal/v1/allocations/requests.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	SortKey string `q:"sort_key"`
+}
+
+// ToAllocationListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToAllocationListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	return q.String(), err
+}
+
+// List makes a request against the API to list ports accessible to you.
```
Comment (jtopjian) [inline openstack/baremetal/v1/allocations/requests.go]: nit: `s/ports/allocations` ?

File: openstack/baremetal/v1/allocations/requests.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			return pagination.Pager{Err: err}
+		}
+		url += query
+	}
+	return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
+		return AllocationPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// Get requests the details off an allocation by ID.
```
Comment (jtopjian) [inline openstack/baremetal/v1/allocations/requests.go]: nit: `s/off/of`

**Comment by stbenjam** [comment]
@jtopjian Updated, should be good to go

**Comment by stbenjam** [comment]
Thank you!

**Pull request merged**